service-clients-core/src/main/java/com/flipkart/poseidon/serviceclients/ServiceContext.java

```diff
@@ -46,40 +44,56 @@ protected Boolean initialValue() {
     private static final ThreadLocal<Boolean> isDebug = ThreadLocal.withInitial(() -> false);
 
     private static final ThreadLocal<Map<String, List<ServiceDebug>>> debugResponses = ThreadLocal.withInitial(ConcurrentHashMap::new);
+    private static ThreadLocal<Map<String, Queue<String>>> collectedHeaders = ThreadLocal.withInitial(HashMap::new);
 
     /**
      * initialize an empty service context, it will cleanup previous value of the threadlocal if used in a threadpool
+     * @param responseHeadersToCollect
      */
-    public static void initialize() {
+    public static void initialize(List<String> responseHeadersToCollect) {
         if (isImmutable.get()) {
             throw new UnsupportedOperationException();
         }
 
         context.remove();
         isImmutable.set(false);
+
+        if (responseHeadersToCollect.isEmpty()) {
+            return;
+        }
+
+        collectedHeaders = ThreadLocal.withInitial(() -> {
```

Comment:
We can still make this final and initialize in static declaration. But here, we can just get map from thread local and add to it (as the map itself isn't immutable)